Api integration

shernanm opened the pull request:
## Summary

* The consumption of the JSONPlaceholder API was implemented.
* The routes for the GET method were added to obtain the albums.

## Trello Card

https://trello.com/c/UsexPSYg

Commit:
API consumption of albums is added

Commit:
Added code for a bad request

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+const axios = require('axios');
+const logger = require('../../app/logger');
```
In Wolox, we defined good practices that for order to requires : first, external dependencies, second, internal dependencies, and finally configs and constants. Use white line space for separte this.

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+const axios = require('axios');
+const logger = require('../../app/logger');
+const PREVIOUS_MESSAGE = 'Making request..';
+const MESSAGE_OK = 'Request made correctly, sending ...';
```
This messages seems like generic .. In this case, if tomorrow there is another controller and it needs to log this, you must to create in the new controller. Solution : Create a file `constants.js` and define all the constants generic.

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+const MESSAGE_OK = 'Request made correctly, sending ...';
+
+const getAlbums = async (req, res) => {
+  const urlAlbums = 'https://jsonplaceholder.typicode.com/albums';
```
Use env var for this. Check config file.

raparicio6 replied:
We put urls and secrets always as env vars

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+const PREVIOUS_MESSAGE = 'Making request..';
+const MESSAGE_OK = 'Request made correctly, sending ...';
+
+const getAlbums = async (req, res) => {
```
Use `(req, res, next)`

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+    logger.info(MESSAGE_OK);
+    res.status(200).send(data);
+  } catch (error) {
+    res.send(error.message);
```
In general, we throw our errors and do not throw the same API error. Create specific error for API fail

raparicio6 replied:
Check how to use the middleware error function

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+
+  try {
+    logger.info(PREVIOUS_MESSAGE);
+    const { data } = await axios.get(urlAlbums);
```
In general, we use promises. It's not bad use it but refactor this :)

molinajulian commented on app/routes.js:
```diff
@@ -1,7 +1,10 @@
 exports.init = app => {
   app.get('/health', healthCheck);
+  app.get('/albums', getAlbums);
+  app.get('/albums/:id/photos', getPhotos);
```
Delete all coments

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+
+const getPhotos = async (req, res) => {
+  try {
+    if (req.params.id > 10 || isNaN(req.params.id)) {
```
For this validations use middlewares. Check package `express-validator`

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+    }
+    logger.info(PREVIOUS_MESSAGE);
+    const urlPhotos = `https://jsonplaceholder.typicode.com/photos?albumId=${req.params.id}`;
+    const { data } = await axios.get(urlPhotos);
```
Data? More specific please

molinajulian replied:
And use a services for communicate with API photos. We use package `request-promise`

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+  }
+};
+
+const getPhotos = async (req, res) => {
```
Same comment in line 6

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,36 @@
+  }
+};
+
+module.exports = { getAlbums, getPhotos };
```
It's no bad but use exports for all functions. For example 
```exports.getYourMoney = () => { return 'I have no money'}```

molinajulian reviewed (changes requested):
Missing test for funcitonalities

Commit:
Changes requested to API consumption

molinajulian commented on app/constants.js:
```diff
@@ -0,0 +1,3 @@
+exports.PREVIOUS_MESSAGE = 'Making request';
+exports.MESSAGE_OK = 'Request made correctly, sending...';
+exports.MESSAGE_API_FAILED = 'Failed to consume the album API';
```
`MESSAGE_API_ALBUM_FAILED`

raparicio6 replied:
Always we try to use descriptive names to be clearer to other programmers who see our code

molinajulian commented on config/index.js:
```diff
@@ -41,6 +41,10 @@ const config = {
       secret: process.env.NODE_API_SESSION_SECRET
+    },
+    resources: {
+      album_url: process.env.URL_ALBUMS,
```
Very good !

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+const albumsService = require('../services/albums'),
+  logger = require('../logger'),
```
Use const for every require

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+
+exports.getAlbumsInventory = (req, res, next) => {
+  logger.info(`${message.PREVIOUS_MESSAGE} to list of albums`);
+  albumsService
```
Return promise

raparicio6 replied:
We always have to put a `return` when we call a promise, to tell Node to "wait" for the promise to resolve/reject and not to get out of the currently function. If we don't put the `return` the endpoint will finish not considering the result of the promise we called

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+    throw Error('Make sure your request is correct');
+  }
+  logger.info(`${message.PREVIOUS_MESSAGE} to list of images of an album by the id: ${idAlbum}`);
+  albumsService
```
Same comment in line 7

molinajulian commented on app/errors.js:
```diff
@@ -8,3 +8,6 @@ exports.databaseError = message => internalError(message, exports.DATABASE_ERROR
 exports.DEFAULT_ERROR = 'default_error';
 exports.defaultError = message => internalError(message, exports.DEFAULT_ERROR);
+
+exports.API_ERROR = 'error_consuming_api';
```
What API? You must be descriptive

molinajulian commented on app/routes.js:
```diff
@@ -1,8 +1,8 @@
-  // app.get('/endpoint/get/path', [], controller.methodGET);
-  // app.put('/endpoint/put/path', [], controller.methodPUT);
-  // app.post('/endpoint/post/path', [], controller.methodPOST);
+  app.get('/albums', getAlbumsInventory);
```
getAlbums its fine

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+exports.getPhotos = (req, res, next) => {
+  const idAlbum = req.params.id;
+  if (idAlbum > 100 || isNaN(idAlbum)) {
+    throw Error('Make sure your request is correct');
```
Maybe throw a errors bad request (status code: 400) and more descriptive too. What is the error? What is wrong in the request?

raparicio6 replied:
Call the error middleware function when there is an error. 
https://expressjs.com/en/guide/error-handling.html

molinajulian commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+  }
+  logger.info(`${message.PREVIOUS_MESSAGE} to list of images of an album by the id: ${idAlbum}`);
+  albumsService
+    .getPhotosById(idAlbum)
```
What pass if tomorrow you need to search photos by name? You can create a function `getPhotosBy` and pass the condition. In this case, the condition is id

molinajulian commented on app/routes.js:
```diff
@@ -1,8 +1,8 @@
-// const controller = require('./controllers/controller');
-const { healthCheck } = require('./controllers/healthCheck');
+const { healthCheck } = require('./controllers/healthCheck'),
+  { getAlbumsInventory, getPhotos } = require('./controllers/album');
```
Add const

raparicio6 commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+    .getAlbums()
+    .then(response => {
+      logger.info(message.MESSAGE_OK);
+      res.status(200).send(response);
```
Here is not necessary to add a `return` because is the last line of the function, but always add it to make the code clearer and to avoid future errors. Another programmer can make more code in this function.

raparicio6 commented on app/controllers/album.js:
```diff
@@ -0,0 +1,30 @@
+  albumsService
+    .getPhotosById(idAlbum)
+    .then(response => {
+      const AlbumImgUrl = response.map(({ url }) => url);
```
variables and functions in camel case

Commit:
Changes requested to API consumption

raparicio6 commented on app/controllers/album.js:
```diff
@@ -1,30 +1,31 @@
   const idAlbum = req.params.id;
   if (idAlbum > 100 || isNaN(idAlbum)) {
-    throw Error('Make sure your request is correct');
+    return next(errors.invalidParameters('Please enter a valid ID'));
```
:clap:

raparicio6 commented on app/services/albums.js:
```diff
@@ -16,10 +16,10 @@ exports.getAlbums = () => {
   const options = {
     method: 'GET',
-    uri: resources.photos_url + id,
+    uri: resources.photos_url + parameter,
```
fix the route using params

Commit:
Changes requested to API consumption

Changes were made in the way to save the URL of the api of the albums and the endpoints. In addition, the property qs was added to the object of the getPhotosBy service for handling the parameters in the requests.

Commit:
The getPhotosBy service is modified to allow more than one filter

raparicio6 commented on app/services/albums.js:
```diff
@@ -0,0 +1,29 @@
+  );
+};
+
+exports.getPhotosBy = qs => {
```
awesome!

molinajulian reviewed (approved):
Fantastic!
